src/gui/GUIManager.ts

```diff
@@ -0,0 +1,676 @@
+                await this.owner.mjolnir.client.sendEvent(this.interactionRoomId, "m.reaction", {
+                    "m.relates_to": {
+                        "rel_type": "m.annotation",
+                        "event_id": requestConfirmationEventId,
+                        "key": `🆗 ${action.emoji} ${await action.title(this.owner, initialNoticeReport)} [${action.label}][${CONFIRM}]`
+                    }
+                });
+                await this.owner.mjolnir.client.sendEvent(this.interactionRoomId, "m.reaction", {
+                    "m.relates_to": {
+                        "rel_type": "m.annotation",
+                        "event_id": requestConfirmationEventId,
+                        "key": `⬛ Cancel [${action.label}][${CANCEL}]`
+                    }
+                });
+            } else {
+                // Execute immediately.
+                LogService.info("GUIManager::handleReaction", "User", event["sender"], "executed (no confirmation needed) action", matches[1]);
+                this.executeAction({
+                    label,
+                    report: confirmationReport,
+                    successEventId: relation.event_id,
+                    failureEventId: relation.eventId,
+                    interactionRoomId: roomId
+                })
+            }
+        }
+    }
+
+
+    /**
+      * Execute a report-specific action.
+      *
+      * This is executed when the user clicks on an action to execute (if the action
+      * does not need confirmation) or when the user clicks on "confirm" in a confirmation
+      * (otherwise).
+      *
+      * @param label The type of action to execute, e.g. `kick-user`.
+      * @param report The abuse report on which to take action.
+      * @param successEventId The event to annotate with a "OK" in case of success.
+      * @param failureEventId The event to annotate with a "FAIL" in case of failure.
+      * @param onSuccessRemoveEventId Optionally, an event to remove in case of success (e.g. the confirmation dialog).
+      */
+    private async executeAction({ label, report, successEventId, failureEventId, onSuccessRemoveEventId, interactionRoomId }: { label: string, report: IReportWithAction, successEventId: string, failureEventId: string, onSuccessRemoveEventId?: string, interactionRoomId: string }) {
+        let action: IUIAction<Owner> | undefined = this.actions.get(label);
+        if (!action) {
+            return;
+        }
+        let error: any = null;
+        let response;
+        try {
+            // Check security.
+            if (interactionRoomId === this.interactionRoomId) {
+                // Always accept actions executed from the interaction room.
+            } else {
+                throw new Error("Security error: Cannot execute this action.");
+            }
+            response = await action.execute(this.owner, report, interactionRoomId, this);
+        } catch (ex) {
+            error = ex;
```

Comment:
do we ever log this exception or is the thinking that the Action should do that? Might be a little risky leaving it to the action